src/modules/image.cxx

```diff
@@ -99,25 +95,51 @@ export class Image
     set_pixels(const std::vector<uint8_t>& pixels)
     {
         this->pixels.clear();
-        this->pixels.reserve(pixels.size() / 3);
-        for (int i = 0; i < pixels.size(); i += 3)
-            this->pixels.push_back(Pixel(pixels[i], pixels[i + 1], pixels[i + 2]));
+        this->pixels.reserve((pixels.size() + channels - 1) / channels);
+        for (int i = 0; i < pixels.size(); i += channels){
+            Pixel<Value_type, channels> p;
+            for (int j = 0; j < channels; j++) {
+                p.values.push_back(pixels[i + j]);
+            }
+            this->pixels.push_back(p);
+        }
     }
 
+    //1 for pbm, 2 for pgm, 3 for ppm
+    template<uint8_t filetype>
     void
-    save_as_ppm(std::string path) const
+    save_as_file(std::string path) const
     {
         std::filesystem::path p = path;
-        std::ofstream ppm_file(path);
-        if (!ppm_file.is_open()){
+        std::ofstream file(path);
+        if (!file.is_open()){
             throw std::runtime_error("Could not open file " + path);
         }
         else {
-            ppm_file  << "P3" << "\n" << std::to_string(width) << " " << std::to_string(height) << "\n" << std::to_string(max_value) << "\n";
-                for (int i = 0; i < width * height; i++)
-                {
-                        ppm_file << pixels[i] << "\n";
-                }
+            file << "P" << std::to_string(filetype) << std::endl;
+            file << std::to_string(width) << " " << std::to_string(height) << std::endl;
+            if constexpr (filetype > 1){
+                file << std::to_string(max_value) << std::endl;
+            }
+            for (int i = 0; i < width * height; i++)
+            {
+                    file << pixels[i] << std::endl;
+            }
+        }
+    }
+
+    template<uint8_t filetype>
+    void
+    print_as_file() const
```

Comment:
This function is not very useful. Maybe write to an ostream, but I'd just have export to file.

Reply:
Oh, this was just to help me debug. I'll remove it.